Sources/ClientGtk/ServerListView.swift

```diff
@@ -41,61 +46,135 @@ struct ServerListView: View {
-            }
-
-            Button("Connect") {
-              completionHandler(server)
-            }
-          case .error(let message):
+          case .server(let index):
+            serverView(index)
+          case .adding:
+            addingView()
+          case .editing(let index):
+            editingView(index)
+          case .error(let message, let returnState):
             Text("Error: \(message)")
-            Button ("Back") {
-              state.detailState = .addition
+            if let returnState {
+              Button ("Back") {
+                state.detailState = returnState
+              }
             }
         }
       }
       .padding(.leading, 10)
     }
   }
 
-  var additionView: some View {
-    VStack {
+  func serverView(_ index: Int) -> some View {
+    let server = state.servers[index]
+
+    return VStack {
+      Text(server.name)
+      if let port = server.port {
+        Text("\(server.host):\(port)")
+      } else {
+        Text(server.host)
+      }
+
+      Button("Connect") {
+        completionHandler(server)
+      }
+
+      Button("Edit") {
+        state.name = server.name
+        if let port = server.port {
+          state.address = "\(server.host):\(port)"
+        } else {
+          state.address = "\(server.host)"
+        }
+        state.detailState = .editing(index)
+      }
+    }
+  }
+
+  func addingView() -> some View {
+    return VStack {
       Text("Add server")
       TextField("Name", state.$name)
       TextField("Address", state.$address)
       
       Button("Add") {
-        let parts = state.address.split(separator: ":")
-        guard parts.count > 0, parts.count <= 2 else {
-          state.detailState = .error("Invalid address")
-          return
+        do {
+          let (host, port) = try ServerListView.parseAddress(state.address)
+
+          let descriptor = ServerDescriptor(name: state.name, host: host, port: port)
+          state.servers.append(descriptor)
+          state.detailState = .server(state.servers.count - 1)
+          
+          state.name = ""
+          state.address = ""
+        } catch AddressParsingError.invalidPort {
+          state.detailState = .error("Invalid port", returnState: .adding)
+        } catch {
+          state.detailState = .error("Invalid address", returnState: .adding)
         }
+      }
+    }
+  }
 
-        let host = String(parts[0])
+  func editingView(_ index: Int) -> some View {
+    return VStack {
+      Text("Edit server")
+      TextField("Name", state.$name)
+      TextField("Address", state.$address)
 
-        var port: UInt16?
-        if parts.count == 2 {
-          guard let parsed = UInt16(parts[1]) else {
-            state.detailState = .error("Invalid port")
-            return
-          }
-          port = parsed
+      Button("Apply") {
+        do {
+          let (host, port) = try ServerListView.parseAddress(state.address)
+
+          state.servers[index].name = state.name
```

Comment:
You can do `state.servers[index] = ServerDescriptor(name: state.name, host: host, port: port)` instead to be more concise